proposal/js-deprecations.d.ts

```diff
@@ -0,0 +1,302 @@
+ * add their own, implementation-specific deprecations.
+ *
+ * Every `Deprecation` has a unique `id`, one of four `status` values, and
+ * (optionally) a human-readable `description`. Depending on the status, each
+ * deprecation may also have a `deprecatedIn` version and an `obsoleteIn`
+ * version that specify the compiler versions the deprecation became active
+ * and became obsolete in, respectively.
+ *
+ * ### Design Decisions
+ *
+ * #### Exposing the Full `Deprecation` Interface
+ *
+ * One alternative to specifying a full `Deprecation` interface is to just have
+ * the relevant APIs take in string IDs. We considered this, but concluded that
+ * each deprecation has additional metadata that users of the API may wish to
+ * access (for example, a bundler may wish to surface the `description` and
+ * `deprecatedIn` version to its users).
+ *
+ * #### Formally Specifying the Deprecations
+ *
+ * We chose to make the list of deprecations part of the specification itself,
+ * as this ensures that the language-wide deprecations are consistent across
+ * implementations. However, if an implementation wishes to add a deprecation
+ * that applies only to itself, it may still do so.
+ *
+ * Additionally, while a deprecation's status is part of the specification, we
+ * chose to leave the `deprecatedIn` and `obsoleteIn` versions of each
+ * deprecation out of the specification. As the two current implementers of this
+ * API are both based on Dart Sass, these versions are _currently_ consistent
+ * across implementations in practice, potential future implementers should not
+ * need to be tied to Dart Sass's versioning.
+ *
+ * #### Errors and Warnings for Invalid Deprecations
+ *
+ * For all of the deprecation options, we chose to have the API throw an error
+ * if a string ID is passed for a deprecation that does not exist. This will
+ * ensure that typos are caught early and is consistent with the behavior on the
+ * command line.
+ *
+ * For `fatalDeprecations`, we chose to make passing in a future deprecation
+ * that's not also passed to `futureDeprecations` an error, since, in a
+ * sense, future deprecations aren't really deprecations yet unless they're
+ * enabled via `futureDeprecations`. This also prevents someone who just passes
+ * all deprecations to `fatalDeprecations` from accidentally opting into a
+ * deprecation that hasn't been released yet. This is not yet consistent with
+ * the behavior on the command line / Dart API, but we plan to update them to
+ * match this specification.
+ *
+ * For obsolete deprecations passed to `fatalDeprecations`, we chose to emit a
+ * warning so users can know to update their configuration, but we allow
+ * compilation to proceed without an error, since an obsolete deprecation means
+ * the underlying behavior has now been removed, which should match the behavior
+ * of making that deprecation fatal while it was still active.
```

Comment:
Is the expectation that deprecations will remain in place even after a major version release? (I think "yes" makes sense here because the nature of deprecations means that different implementations may remove them at different times, but this is worth documenting I think.)

Reply:
Yes, that was my intention with adding the concept of obsolete deprecations. Added clarification for this above.